### PR TITLE
Add support for latexmkrc

### LINKED ISSUE #179: Add support for makeglossaries
It would be great if there was a setting to allow the user to add extra commands during compilation, for instance the `makeglossaries` command.

### PR DESCRIPTION
**Pull request by yitzchak**
This PR adds support for our own latexmkrc file. Currently supports Asymptote, `glossaries`, MetaPost, `nomencl`, `index` and SageTeX. Resolves #179.

### Current support and tests

- [x] `-r` spec
- [x] Asymptote
  - [x] spec
  - [x] CI support
- [x] `glossaries` package
  - [x] spec
  - [x] CI support
- [x] `mpost`
  - [x] spec
  - [x] CI support
- [x] `nomencl` package
  - [x] spec
  - [x] CI support
- [x] `index` package
  - [x] spec
  - [x] CI support
- [x] SageTeX
  - [x] spec
  - [x] CI support (going to wait on this since it is 1GB in size)

### CODE REVIEW

**Comment by thomasjo** [comment]
Looks really promising! This will need some tests ✅

File: spec/builders/latexmk-spec.js [hunk 1]
```diff
@@ -214,5 +222,52 @@ describe('LatexmkBuilder', () => {
...
+
+        waitsForPromise(() => {
+          return builder.run(filePath).then(code => { exitCode = code })
+        })
+
+        runs(() => {
+          expect(exitCode).toBe(0)
+          for (const ext of extensions) {
+            const output = path.format({ dir, name, ext })
+            expect(fs.existsSync(output)).toBe(true, `Check the existance of ${ext} file.`)
```
Comment (thomasjo) [inline spec/builders/latexmk-spec.js]: existence

File: spec/builders/latexmk-spec.js [hunk 2]
```diff
@@ -214,5 +222,52 @@ describe('LatexmkBuilder', () => {
...
+
+    checkCustomRules('successfully creates index files when using the index package',
+      'index-test',
+      ['.idx', '.ind', '.ldx', '.lnd', '.pdf']
+    )
+
+    checkCustomRules('successfully creates SageTeX files when using the sagetex package',
+      'sage-test',
+      ['.sagetex.sage', '.sagetex.sout', '.pdf']
+    )
```
Comment (thomasjo) [inline spec/builders/latexmk-spec.js]: Is there some way we can avoid all this "custom" stuff? It makes the tests really hard to read. Let's stick to more idiomatic Jasmine code. Basically I don't want to see any `it` calls inside other functions or in loops, etc.

I'd much rather have code duplication in specs, than lots of nested function calls etc which makes it hard to read the specs. If I have to "parse" the code to figure out what tests will be run, there's a problem.

File: package.json [hunk 3]
```diff
@@ -117,6 +117,13 @@
       "default": true,
       "order": 6
     },
+    "enableLatexmkrc": {
```
Comment (thomasjo) [inline package.json]: I'm starting to think we should call this something else. Is it really important to the end user that this functionality is provided via a custom `.latexmkrc` file? That seems like an implementation detail leaking out. We should name this based on the functionality gained by enabling this option.

Comment (yitzchak) [inline package.json]: How about `enableExtendedRules` or `enableCustomRules`?

Comment (thomasjo) [inline package.json]: I like those, but what about something even more verbose, e.g. `enableExtendedBuildRules`? Perhaps even avoiding the somewhat ambiguous "rules" term with something like `enableExtendedBuildMode` or `enableExtendedBuild`?

Comment (yitzchak) [inline package.json]: `enableExtendedBuildMode` sounds good!

**Comment by yitzchak** [comment]
@thomasjo Looks like we need at least the `index`, `glossaries` and `nomencl` packages added to the TeX Live CI images. That include `makeindex` and `makeglossaries` scripts.

**Comment by yitzchak** [comment]
@thomasjo Added Asymptote support so we will need that also.

**Comment by thomasjo** [comment]
I just checked and the requested packages are already installed in the TeX Live images. In other words, `asymptote`, `glossaries`, `index`, and `nomencl` are already available.

**Comment by yitzchak** [comment]
Thanks, must be some other problem with the specs.

**Comment by thomasjo** [comment]
No wait, I accidentally ran `tlmgr list --only-installed` instead of `./tlmgr list --only-installed`...
Only `asymptote` is installed. I'm updating the the distributions now! Sorry about the confusion 🙇

**Comment by yitzchak** [comment]
No worries! I just noticed that with the console debug messages that I just added.

**Comment by yitzchak** [comment]
I added `feynmp` for a mpost test btw.

**Comment by thomasjo** [comment]
Do you mean [`feynmf`](https://www.ctan.org/pkg/feynmf)? As far as I can tell, there's no longer any TeX Live package called `feynmp`.

**Comment by yitzchak** [comment]
Yes, they are packaged together. Sorry.

**Comment by yitzchak** [comment]
`index` and `nomencl` look good now. Need `xkeyval` and `mpost` for the `glossaries`/`feynmf` test.

**Comment by thomasjo** [comment]
Try now; just added `metapost` and `xkeyval` to the travis-texlive distribution.

**Comment by yitzchak** [comment]
That fixed the mpost test. Glossaries is complaining about `mfirstuc`. The total list of dependencies for glossaries looks to be: `mfirstuc`, `textcase`, `xfor`, `datatool-base`, `amsgen` (probably already there since it is in AMSLaTeX) and `etoolbox`.

**Comment by thomasjo** [comment]
Bumped the distribution again. `amsgen.sty` is packaged with the `amsmath` package as you expected.

**Comment by yitzchak** [comment]
It's complaining about `substr`. I looked in the the FDB and this is probably what is left: `substr`, `supertabular`, `longtable`.

Wish I had thought of that sooner. Sorry.

**Comment by yitzchak** [comment]
Also, I am not sure we will be able to test Asymptote on Travis. It is trying to use OpenGL to do rendering.

**Comment by thomasjo** [comment]
You should be able to fix that by using a simpler fixture — stick to 2D. The problem is `import three` relies on OpenGL by default. Might be possible to tweak, but just tweak the fixture.

**Comment by thomasjo** [comment]
Bumped the distribution again. Added `substr` and `supertabular`.

**Comment by yitzchak** [comment]
Asymptote is fixed! Good catch. Looks like we need `fp`. The FDB appears to be in no particular order.

```
  "/usr/local/texlive/2016/texmf-dist/fonts/type1/public/amsfonts/cm/cmr10.pfb" 1248133631 35752 024fb6c41858982481f6968b5fc26508 ""
  "/usr/local/texlive/2016/texmf-dist/tex/generic/xkeyval/keyval.tex" 1403829539 2725 fc34ef3ccb37ba15a640e8fca6190bca ""
  "/usr/local/texlive/2016/texmf-dist/tex/generic/xkeyval/xkeyval.tex" 1417732693 19231 26434a5656c684f5ffb1f26f98006baa ""
  "/usr/local/texlive/2016/texmf-dist/tex/generic/xkeyval/xkvutils.tex" 1403829539 7677 6f5ce7c1124cad7ec57d05b2562bd8fe ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/amsmath/amsbsy.sty" 1456875012 2210 5c54ab129b848a5071554186d0168766 ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/amsmath/amsgen.sty" 1456875012 4160 c115536cf8d4ff25aa8c1c9bc4ecb79a ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/amsmath/amsmath.sty" 1467147727 82033 ff9a037b5bad3f86db2c1a9e5c8160c3 ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/amsmath/amsopn.sty" 1459635588 4115 318a66090112f3aa3f415aeb6fe8540f ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/amsmath/amstext.sty" 1456875012 2431 fe3078ec12fc30287f568596f8e0b948 ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/base/article.cls" 1459635588 19821 310da678527a7dfe2a02c88af38079b7 ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/base/ifthen.sty" 1454284088 5159 a08c9bbd48fc492f15b22e458bef961f ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/base/size10.clo" 1459635588 8292 e897c12e1e886ce77fe26afc5d470886 ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/datatool/datatool-base.sty" 1469920206 87642 332d6d66c1554a262967ad743548804a ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/datatool/datatool-fp.sty" 1469920206 3944 ab54cce925388bd946e8068230bbe5dd ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/etoolbox/etoolbox.sty" 1438639395 41547 7860e623236f5f726a4099bbb7dd89fb ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/fp/defpattern.sty" 1137110169 1081 6a8d78d0e6b2d89334f92fd122c1da99 ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/fp/fp-addons.sty" 1137110169 1275 fe903646b72b3e877914d78f2db9ddda ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/fp/fp-basic.sty" 1137110169 29695 b6366b50463dfb367cf560419d0cea8d ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/fp/fp-eqn.sty" 1137110169 11336 afed7aa5dbc4bf246950282b95e522c2 ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/fp/fp-eval.sty" 1137110169 6220 eeeffd00d1235e73c380c6dc9f30ac43 ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/fp/fp-exp.sty" 1137110169 16964 7ba2d4a3882e46a0429106267976ad21 ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/fp/fp-pas.sty" 1137110169 3006 6fd5da0dc83ba1d19525f1fb08dca7df ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/fp/fp-random.sty" 1137110169 3702 6e54ac63212901d59e6ea2a32b0a7689 ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/fp/fp-snap.sty" 1137110169 6519 c1e5d3c6aed32ba782f756d79578d87b ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/fp/fp-trigo.sty" 1137110169 28969 342a2de338e342de95d2551557e50f06 ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/fp/fp-upn.sty" 1137110169 7154 78a835a719e27a484e63f1f98ba4df11 ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/fp/fp.sty" 1137110169 983 d84df22b060ce125f89a32c403dcab31 ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/glossaries/base/glossaries-compatible-307.sty" 1476572210 18527 55cc3270bfbf26944be2c8ed7f041025 ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/glossaries/base/glossaries.sty" 1476572210 236070 e3c50ed560cc1ac2f152d244ddbd9ddf ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/glossaries/styles/glossary-hypernav.sty" 1476572210 5395 7f94dd30bdae46966c0e7ca5a230df07 ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/glossaries/styles/glossary-list.sty" 1476572210 6443 5b760430f42fc86dd20f5257bb0051c2 ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/glossaries/styles/glossary-long.sty" 1476572210 8857 f87680122b872acabd9f258190b71bed ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/glossaries/styles/glossary-super.sty" 1476572210 9874 e4b71cb855734f075e41b394c79e6fe7 ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/glossaries/styles/glossary-tree.sty" 1476572210 12339 bededa26535bf01196dad9c8e82db944 ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/mfirstuc/mfirstuc.sty" 1470088054 11863 cf40ea201f88e5381ca4cc5e5dc5cecb ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/substr/substr.sty" 1258848659 5724 d3505925e87b345f11a5f17d58550dea ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/supertabular/supertabular.sty" 1137110892 14283 563cabe54cdab0459534c65c66672d04 ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/textcase/textcase.sty" 1137110938 2022 79a2c3efd954daff1698214872b6ca46 ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/tools/longtable.sty" 1454284088 12083 80916157594a8e4354985aaefae4f367 ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/xfor/xfor.sty" 1233902931 2689 44855b72da58344b1c1d9b3087bce6ca ""
  "/usr/local/texlive/2016/texmf-dist/tex/latex/xkeyval/xkeyval.sty" 1417732693 4962 9c1069474ff71dbc47d5006555e352d3 ""
  "/usr/local/texlive/2016/texmf-dist/web2c/texmf.cnf" 1470266149 31958 926ad82bc47ee89401cf24be7d2cbf7d ""
  "/usr/local/texlive/2016/texmf-var/fonts/map/pdftex/updmap/pdftex.map" 1476747522 1950563 ac64466290d324653f4ce0242627de5f ""
  "/usr/local/texlive/2016/texmf-var/web2c/pdftex/pdflatex.fmt" 1473012305 3873676 ab218122bbfd38b731a8dd5c1a6aa3a9 ""
  "/usr/local/texlive/2016/texmf.cnf" 1465329017 1019 001384a96e7d401aefbc605835295e19 ""
```

**Comment by yitzchak** [comment]
@thomasjo Specs are now functional in CI. I think one of the last remaining issues is whether we want this to be on by default or not. I think that there are three options:

1. Leave it off by default. Protects the latexmkrc configuration of current users but is the least opinionated choice.
2. Turn it on by default. Since the command line option `-r` is read after latemkrc system and home directories are scanned our file will overwrite user's rules (only for the same extensions) unless the user turns it off.
3. Turn it on by default and pass the `-norc` option to prevent system and user files from being loaded. Then in our RC file add the following at the end. This would essentially make our file the first system RC file loaded.

```perl
# System rc file:
read_first_rc_file_in_list( @rc_system_files );
}
if ($HOME ne "") {    
    # User rc file:
    # N.B. $HOME equals "" if latexmk couldn't determine a home directory.
    # In that case, we shouldn't look for an rc file there.
    read_first_rc_file_in_list( "$HOME/.latexmkrc" );
}
# Rc file in current directory:
read_first_rc_file_in_list( "latexmkrc", ".latexmkrc" );

```

**Comment by yitzchak** [comment]
@thomasjo I am leaning toward option 2, i.e. turn on extended rules by default.

**Comment by thomasjo** [comment]
What about using an enum instead of bool? With options like `off`, `inclusive`, `exclusive`. I realize the names there are not very descriptive, but the idea is to support all three alternatives you've outlined above. Regardless of what we names the options (or whether we use a bool), we will have to document this quite thoroughly. A link to a wiki page from the README is probably good enough, combined with a somewhat brief description in the config schema.

If we go with the enumeration approach, I'm comfortable with setting the default to `exclusive` (or whatever we call it), since we then give the user an option to change it to an inclusive mode in case they have some custom rules they want/need, while still guaranteeing a baseline experience out of the box for the average user.

**Comment by yitzchak** [comment]
I am fine with an enumeration, although a fourth possible has occurred to me, which is the system ones, then ours, then the users.

**Pull request merged**